File: botleague-backend/src/main/java/com/botleague/backend/events/service/MatchFormatPolicy.java
package com.botleague.backend.events.service;

import com.botleague.backend.events.enums.MatchFormatKind;

/**
 * The single source of truth for "which match-generation system this sport uses" —
 * elimination bracket (combat/team sports) vs. round-wise time trial (timed-run sports).
 *
 * <pre>
 *   Sport            Format
 *   Robo War         Bracket
 *   Robo Sumo        Bracket
 *   Drone Soccer     Bracket
 *   Robo Soccer      Bracket
 *   Robo Race        Round Time Trial
 *   RC Racing Car    Round Time Trial
 *   Line Follower    Round Time Trial
 * </pre>
 *
 * Unlike SportSpecPolicy, this doesn't vary by league/age-group — the same sport
 * always uses the same match format regardless of which league it's run under.
 * A sport that isn't in the table falls back to BRACKET, the existing default
 * behavior, so an unrecognized/custom sport never silently loses its bracket.
 */
public final class MatchFormatPolicy {

    private MatchFormatPolicy() {}

    public static MatchFormatKind formatFor(String sportName) {
        switch (sportKey(sportName)) {
            case "ROBORACE":
            case "RCRACINGCAR":
            case "LINEFOLLOWER":
                return MatchFormatKind.ROUND_TIME_TRIAL;
            default:
                return MatchFormatKind.BRACKET;
        }
    }

    /** Shared with SportSpecPolicy and the registration sport check — see SportKeys. */
    private static String sportKey(String sportName) {
        return SportKeys.of(sportName);
    }
}

File: Botleague-Frontend/src/feature/Event/utils/matchFormatPolicy.ts
// Mirror of backend MatchFormatPolicy — which match-generation system a sport
// uses: the existing elimination bracket (combat/team sports) or the newer
// round-wise time trial (timed-run sports). See specPolicy.ts for the sibling
// weight/dimension policy this file's normalization approach is copied from.
//
//   Sport            Format
//   Robo War         BRACKET
//   Robo Sumo        BRACKET
//   Drone Soccer     BRACKET
//   Robo Soccer      BRACKET
//   Robo Race        ROUND_TIME_TRIAL
//   RC Racing Car    ROUND_TIME_TRIAL
//   Line Follower    ROUND_TIME_TRIAL

export type MatchFormatKind = "BRACKET" | "ROUND_TIME_TRIAL";

const norm = (s?: string | null) => (s ?? "").toUpperCase().replace(/[^A-Z0-9]+/g, "");

function sportKey(sport?: string | null): string {
  const n = norm(sport);
  if (n.includes("ROBOWAR") || n.includes("ROBOTWAR") || n.includes("COMBAT")) return "ROBOWAR";
  if (n.includes("ROBOSUMO") || n.includes("SUMO")) return "ROBOSUMO";
  if (n.includes("LINEFOLLOWER") || n.includes("LINEFOLLOW")) return "LINEFOLLOWER";
  if (n.includes("DRONESOCCER") || n.includes("DRONE")) return "DRONESOCCER";
  if (n.includes("RCRACING") || n.includes("RCROBO") || n.includes("RCCAR")) return "RCRACINGCAR";
  if (n.includes("ROBOSOCCER") || n.includes("SOCCER")) return "ROBOSOCCER";
  if (n.includes("ROBORACE") || n.includes("ROBORACING") || n.includes("RACE")) return "ROBORACE";
  return n;
}

const ROUND_TIME_TRIAL_SPORTS = new Set(["ROBORACE", "RCRACINGCAR", "LINEFOLLOWER"]);

export function formatFor(sport?: string | null): MatchFormatKind {
  return ROUND_TIME_TRIAL_SPORTS.has(sportKey(sport)) ? "ROUND_TIME_TRIAL" : "BRACKET";
}
